File: Protocol/_theoreme_sim/src/state/simulationStep.js
/* Purpose: Execute one full simulation round with shared logic for UI and No-UI runners. */

import { maybeRunDecayPhase } from "../phases/decayPhase.js";
import { runDownRoundPhase } from "../phases/downRoundPhase.js";
import { runUpRoundPhase } from "../phases/upRoundPhase.js";
import { finalizeLinkStrengthRound } from "../propagation/linkUsageTracker.js";
import { applyChargeSpreadRound } from "../propagation/chargeSpreadRound.js";
import { propagateNeighborChargesRound } from "../propagation/neighborChargePropagation.js";
import { rebuildTree } from "../routing/treeBuilder.js";
import { updateOscillationReport } from "../verification/oscillationDetector.js";
import { evaluateTheorem } from "../verification/theoremChecks.js";
import { refreshChargeBounds, refreshEligibility } from "./simulationState.js";

/**
 * @param {any} state
 * @returns {{advanced:boolean}}
 */
export function advanceSimulationRound(state) {
  if (!state) {
    return { advanced: false };
  }

  if (state.round >= state.config.maxRounds) {
    return { advanced: false };
  }

  state.round += 1;

  state.lastBroadcastReport = runDownRoundPhase(state);
  state.lastUp = runUpRoundPhase(state);
  state.lastPropagation = propagateNeighborChargesRound(state);
  state.lastSpread = applyChargeSpreadRound(state);

  finalizeLinkStrengthRound(state);

  refreshEligibility(state.nodes, state.config.qForward);

  let rebuilt = rebuildTree(state);
  state.parentMap = rebuilt.parentMap;
  state.childrenMap = rebuilt.childrenMap;
  let finalChangedCount = rebuilt.changedCount;

  state.lastDecay = maybeRunDecayPhase(state);
  if (state.lastDecay.triggered) {
    refreshEligibility(state.nodes, state.config.qForward);
    rebuilt = rebuildTree(state);
    state.parentMap = rebuilt.parentMap;
    state.childrenMap = rebuilt.childrenMap;
    finalChangedCount = rebuilt.changedCount;
  }

  state.stableRounds = finalChangedCount === 0 ? state.stableRounds + 1 : 0;

  refreshChargeBounds(state);
  state.lastOscillationReport = updateOscillationReport(state);
  state.lastTheoremReport = evaluateTheorem(state);

  if (!Array.isArray(state.broadcastHistory)) {
    state.broadcastHistory = [];
  }
  state.broadcastHistory.push({
    round: state.round,
    duplicates: state.lastBroadcastReport.duplicates,
    reachedCount: state.lastBroadcastReport.reachedCount,
    coverage: state.lastBroadcastReport.coverage,
  });

  if (state.broadcastHistory.length > 300) {
    state.broadcastHistory.splice(0, state.broadcastHistory.length - 300);
  }

  return { advanced: true };
}

File: Protocol/_theoreme_sim_py/src/state/simulation_step.py
"""Purpose: Execute one full simulation round in headless mode."""

from __future__ import annotations

from src.phases.decay_phase import maybe_run_decay_phase
from src.phases.down_round_phase import run_down_round_phase
from src.phases.up_round_phase import run_up_round_phase
from src.propagation.charge_spread_round import apply_charge_spread_round
from src.propagation.link_usage_tracker import finalize_link_strength_round
from src.propagation.neighbor_charge_propagation import propagate_neighbor_charges_round
from src.routing.tree_builder import rebuild_tree
from src.state.simulation_state import refresh_charge_bounds, refresh_eligibility
from src.verification.oscillation_detector import update_oscillation_report
from src.verification.theorem_checks import evaluate_theorem


def advance_simulation_round(state: dict) -> dict:
    if state is None:
        return {"advanced": False}

    if int(state["round"]) >= int(state["config"]["maxRounds"]):
        return {"advanced": False}

    state["round"] += 1

    state["lastBroadcastReport"] = run_down_round_phase(state)
    state["lastUp"] = run_up_round_phase(state)
    state["lastPropagation"] = propagate_neighbor_charges_round(state)
    state["lastSpread"] = apply_charge_spread_round(state)

    finalize_link_strength_round(state)

    refresh_eligibility(state["nodes"], state["config"]["qForward"])

    rebuilt = rebuild_tree(state)
    state["parentMap"] = rebuilt["parentMap"]
    state["childrenMap"] = rebuilt["childrenMap"]
    final_changed_count = rebuilt["changedCount"]

    state["lastDecay"] = maybe_run_decay_phase(state)
    if state["lastDecay"]["triggered"]:
        refresh_eligibility(state["nodes"], state["config"]["qForward"])
        rebuilt = rebuild_tree(state)
        state["parentMap"] = rebuilt["parentMap"]
        state["childrenMap"] = rebuilt["childrenMap"]
        final_changed_count = rebuilt["changedCount"]

    state["stableRounds"] = (
        int(state["stableRounds"]) + 1 if final_changed_count == 0 else 0
    )

    refresh_charge_bounds(state)
    state["lastOscillationReport"] = update_oscillation_report(state)
    state["lastTheoremReport"] = evaluate_theorem(state)

    if not isinstance(state.get("broadcastHistory"), list):
        state["broadcastHistory"] = []

    state["broadcastHistory"].append(
        {
            "round": state["round"],
            "duplicates": state["lastBroadcastReport"]["duplicates"],
            "reachedCount": state["lastBroadcastReport"]["reachedCount"],
            "coverage": state["lastBroadcastReport"]["coverage"],
        }
    )

    if len(state["broadcastHistory"]) > 300:
        state["broadcastHistory"] = state["broadcastHistory"][-300:]

    return {"advanced": True}
